nRF Wi-Fi Management

krish2718 opened the pull request:
More extensions for wifi_mgmt, added status and statistics APIs. These are currently used by wpa_supplicant whose changes are still in NCS, will be upstreamed to Zephyr in the future, but are generic to be implemented by any offload drivers. Only a minimal status and statistics are added as a starting point.

Commit:
wifi_mgmt: Implement checks for new security types

Extend checks for PSK-256 and SAE.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_mgmt: Add new API for Wi-Fi status

A new net_mgmt command and event are added for interface status,
depending on the implementation the status can be returned when polled
or an unsolicited event can be send by driver whenever there is a change
in status.

This is planned to be implemented only by upcoming wpa_supplicant,
offload implementation is left for driver developers.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_mgmt: Add support for connection timeout

This is handy for applications to control timeout and try multiple
connections rather than relying on underlying implementation.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_mgmt: Add a comment for legacy types

As this was deliberate add a comment that we support and encourage only
None/WPA2/WPA3.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_mgmt: Add new API for Wi-Fi statistics

Networking statistics framework is used to define handler and the data
structure, Wi-Fi management layer implements the handler and also adds a
new offload API to get statistics from the Wi-Fi driver.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_shell: Rename Wifi to Wi-Fi

This is the standard.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
release_notes: Add new offload APIs

Mention about newly added offload APIs.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

krish2718 commented:
>Error: subsys/net/l2/wifi/wifi_shell.c:269:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell

This seems to be a newly introduced coverity MISRA standards scan, its applicable for existing code as well but I guess its running only on diff, so, didn't report that.  I will fix the newly introduced code, How do we handle this for existing one, a separate commit/PR?

mbolivar-nordic commented:
> How do we handle this for existing one, a separate commit/PR?

I usually do separate commits for such things, FWIW. Example: 93237fcdde01f7477b830277abbe94c3d4f97ae5

mniestroj commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -119,6 +129,20 @@ struct wifi_status {
 };
 
+struct wifi_iface_status {
+	int state; /* enum wifi_iface_state */
```
Why don't we use `enum wifi_iface_state` instead of `int`?

krish2718 replied:
Typically, `int` is used in definition for enum's to avoid ambiguity as some compilers/systems use `char/int` for enums. This is mainly for the interface boundary files (host to chip) as they use different compilers/systems.

In this case, I can use `enum`.

krish2718 replied:
So, two options: 

1. use `int` and return `-1` for unknown
2. use `enum` and return `*_UNKNOWN`

I prefer #1 but okay to switch to #2 . Any hard preferences?

krish2718 replied:
Usage would be `enum`, but storage would be `int` to avoid incompatibility, but I can change to use `enum` for storage as well, it should be okay for now.

mniestroj commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -266,6 +288,51 @@ static int cmd_wifi_scan(const struct shell *shell, size_t argc, char *argv[])
+		shell_fprintf(shell, SHELL_NORMAL, "Channel: %d\n", status.channel);
+		shell_fprintf(shell, SHELL_NORMAL, "Security: %s\n",
+				wifi_security_txt(status.security));
+		shell_fprintf(shell, SHELL_NORMAL, "PMF: %s\n",
```
s/PMF/MFP/ ?

krish2718 replied:
This was intentional, MFP is the IEEE name, but PMF is the most commonly used name, so, using PMF for public.

mniestroj replied:
But this is basically the same thing? Won't it be confusing if we mix those two? Usually it is better to stick with one, so that `grep` does the job without the requirement to have context of what PMF/MFP means (e.g. I don't know what it means).

krish2718 replied:
I was just following wpa_supplicant style (https://w1.fi/wpa_supplicant/devel/structwpa__config.html#adc14d004b51292339ada9ed6f5d276b8), but sure we can stick to MFP.

MFP - Management frame protection
PMF - Protected Management frames

mniestroj replied:
Personally I would prefer to use one or the other, I feel like it would be easier to jump in for newcomers (like me in case of WiFi). But don't treat that as hard requirement from me.

I didn't work much with `wpa_supplicant`, but heard a lot about legacy that comes with it. I think that using one name for a single thing should be the way to go in our codebase. If needed, we can use both names in the glue layer between wpa_supplicant and Zephyr.

mniestroj commented on include/zephyr/net/wifi.h:
```diff
@@ -25,24 +25,197 @@ enum wifi_security_type {
+	case WIFI_7:
+		return "WIFI 7 (802.11be/EHT)";
+	default:
+		return "UNKNOWN";
```
Should we have a dedicated enum value for "UNKNOWN"? I suspect that some WiFi offloaded drivers won't have that information.

krish2718 replied:
We could, I am just using `-1` as the return value in case this is not filled.

krish2718 replied:
Even if we have dedicated enum for UNKNOWN, we still need to handle `default` case, so, are you suggesting to use `WIFI_LINK_MODE_UNKNOWN->"UNKNOWN"` and then `default->"not available"`?

krish2718 replied:
I think I understand now, I have removed `default` and added UNKNOWN and MAX ENUMs for all, please check.

krish2718 replied:
urgh..checkpatch is throwing a warning : `../../../../../../include/zephyr/net/wifi.h:157:1: error: control reaches end of non-void function [-Werror=return-type]`, so, need a `default` case, but what should be the return value `-1`, or just remove `case UNKNOWN`?

mniestroj replied:
What you have done with:
```
	case WIFI_SECURITY_TYPE_UNKNOWN:
	default:
		return "UNKNOWN";
```
look good to me.

My main point was about returning UNKNOWN enum instead of -1 by drivers implementing this interface. Thanks for changes!

mniestroj commented on subsys/net/ip/Kconfig.stats:
```diff
@@ -125,4 +125,12 @@ config NET_STATISTICS_POWER_MANAGEMENT
+	help
+	  Keep track of Wi-Fi related statistics. Note that this
+	  requires support from the Wi-Fi driver. The driver needs
+	  to collect the statistics.
```
Missing newline after help text.

mniestroj commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -288,45 +288,45 @@ static int cmd_wifi_scan(const struct shell *shell, size_t argc, char *argv[])
 }
 
-static int cmd_wifi_status(const struct shell *shell, size_t argc, char *argv[])
+static int cmd_wifi_status(const struct shell *shell_h, size_t argc, char *argv[])
```
I think that `const struct shell *sh` is more commonly used across Zephyr tree.

Commit:
wifi_shell: Fix the duplicate tag CI warning

Newly introduced Coverity scan throws a warning about duplicate tag as
per MISRA coding standards, so, use a unique tag name in the existing
code for "shell".

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

mniestroj commented:
@sylvioalves Could you or someone from Espressif team look at those changes?

Commit:
wifi_mgmt: Extend enumerations to include MAX and UNKNOWN

This is handy to

* Validate the input using MAX
* Return UNKNOWN instead of int "-1"

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_shell: Add new security methods

Add PSK_256, SAE and MFP settings.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
wifi_mgmt: Fix checkpatch warnings

Checkpatch says "else after return is not useful".

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

rlubos reviewed (commented):
Looks pretty good from my side.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -150,13 +150,9 @@ struct net_wifi_mgmt_offload {
  */
 BUILD_ASSERT(offsetof(struct net_wifi_mgmt_offload, iface_api) == 0);
 
-#ifdef CONFIG_WIFI_OFFLOAD
```
Yes, I left a comment there as well, will drop this commit once its merged.

rlubos commented:
It seems this PR can already be rebased - https://github.com/zephyrproject-rtos/zephyr/pull/49464 was merged.

mniestroj commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -157,3 +164,71 @@ static int wifi_ap_disable(uint32_t mgmt_request, struct net_if *iface,
+		struct wifi_iface_status *iface_status)
+{
+	net_mgmt_event_notify_with_info(NET_EVENT_WIFI_IFACE_STATUS,
+					iface, &iface_status,
```
```suggestion
					iface, iface_status,
```

mniestroj commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -157,3 +164,71 @@ static int wifi_ap_disable(uint32_t mgmt_request, struct net_if *iface,
+}
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_IFACE_STATUS, wifi_iface_status);
+
+void wifi_mgmt_raise_iface_status_event(struct net_if *iface,
+		struct wifi_iface_status *iface_status)
```
I am not sure this is needed. First you define `NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_IFACE_STATUS, wifi_iface_status);` with implementation that looks synchronous. Then you add `wifi_mgmt_raise_iface_status_event`, which is (I think) asynchronous response to request. I guess we should go with synchronous or asynchronous, not both. See "connect" and "disconnect" which are asynchronous requests. See "ap_enable" which is synchronous request.

krish2718 replied:
We can support both, e.g., if an application wants to fetch status then it can use the `synchronous` way, and if the underlying driver/wpa_supplicant wants to notify the application of an important event then it can use `asynchronous` way.

mniestroj replied:
Oh, I get what you mean. This will be the first wifi_mgmt/net_mgmt API that does this, right?

krish2718 replied:
Yes.

mniestroj commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -201,3 +201,34 @@ void wifi_mgmt_raise_iface_status_event(struct net_if *iface,
+		return -EINVAL;
+	}
+
+	ret = off_api->get_stats(dev, &stats);
```
How about passing `data` directly instead of stack allocated stats + memcpy?
